Fixed use of ECB with Botan

antoinelochet opened the pull request:
Here is my proposed fix for issue 723

bjosv reviewed (commented):
Seems reasonable.
Are all the extra debug printouts needed? I seems that existing printouts are quite sparse with information, maybe for security reasons(?).

As an addition I suggest that we also disable 2 testcases when built WITH_BOTAN.
They will always fail when using Botan as I understand it. WDYT?

```
1) test: AESTests::testECB (F) line: 518 AESTests.cpp
assertion failed
- Expression: aes->encryptInit(&aesKey128, SymMode::ECB, IV)


2) test: DESTests::testECB (F) line: 537 DESTests.cpp
assertion failed
- Expression: des->encryptInit(&desKey56, SymMode::ECB, IV)
```

antoinelochet commented on src/lib/data_mgr/SecureDataManager.cpp:
```diff
@@ -397,6 +397,7 @@ void SecureDataManager::logout()
 // Decrypt the supplied data
 bool SecureDataManager::decrypt(const ByteString& encrypted, ByteString& plaintext)
 {
+    DEBUG_MSG("encrypted: %s", encrypted.hex_str().c_str());
```
I have deleted the debug messages and reindented all the `SecureDataManager.cpp` file

antoinelochet commented:
I think that you are right about the tests.

I have added the debug logs because they are often useful. I agree that some security reasons, it may seem bad, but:
- using SoftHSM in production is madness
- using SoftHSM in production WITH debug logs is beyond madness
So I thought this was tolerable. But per your request, I deleted all the offending lines.

bjosv commented on src/lib/crypto/test/AESTests.cpp:
```diff
@@ -328,6 +328,7 @@ void AESTests::testCBC()
 	}
 }
 
+#ifndef WITH_BOTAN
```
I believe this will give an undefined reference to this function while linking.
Maybe its better to just remove the registration of the testcase in the header file.

```
#ifndef WITH_BOTAN
        CPPUNIT_TEST(testECB);
#endif
```

bjosv replied:
Same for the `testECB()` testcase in:
src/lib/crypto/test/DESTests.h

antoinelochet replied:
I followed your suggestion and compilation/linking is ok now.
The tests fail on the src/lib/test though but I am pretty sure that it was the case before my changes.

bjosv replied:
Great! Looks good.

bjosv reviewed (commented):
Good comments. Using debug logs in production would be "beyond madness" :)

anpa8480 commented:
if you revert BotanSymmetricAlgorithm.cpp 8fd89ec33edfe4a7b510c471101f78b52a0c1294 and uncomment all tests are passing

antoinelochet commented:
> if you revert BotanSymmetricAlgorithm.cpp [8fd89ec](https://github.com/opendnssec/SoftHSMv2/commit/8fd89ec33edfe4a7b510c471101f78b52a0c1294) and uncomment if (mode == SymMode::ECB) all tests are passing.
> 
> Tested with Botan 2.19.3 ![image](https://private-user-images.githubusercontent.com/173441517/341674688-0cabeb23-97f5-4606-b3b6-2af4c11d6a4e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg5NjIwMzIsIm5iZiI6MTcxODk2MTczMiwicGF0aCI6Ii8xNzM0NDE1MTcvMzQxNjc0Njg4LTBjYWJlYjIzLTk3ZjUtNDYwNi1iM2I2LTJhZjRjMTFkNmE0ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyMVQwOTIyMTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mYmZlMzY0ZDAzZGM4OTdmYTVhMzM0YzE5N2ZkYTU3NjIzZjFmOWUzZWEzNWZkMWI5YjUzNzhlMDY5ODBkODI3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.q73NK29uMcP9pcJkVDJ3WJ20COBRP2WtnfKsVIh9G94)

That's because ECB is not taken into account during compilation phase. It does not fix the failing tests in `src/lib`

anpa8480 commented:
> > if you revert BotanSymmetricAlgorithm.cpp [8fd89ec](https://github.com/opendnssec/SoftHSMv2/commit/8fd89ec33edfe4a7b510c471101f78b52a0c1294) and uncomment if (mode == SymMode::ECB) all tests are passing.
> > Tested with Botan 2.19.3 ![image](https://private-user-images.githubusercontent.com/173441517/341674688-0cabeb23-97f5-4606-b3b6-2af4c11d6a4e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg5NjIwMzIsIm5iZiI6MTcxODk2MTczMiwicGF0aCI6Ii8xNzM0NDE1MTcvMzQxNjc0Njg4LTBjYWJlYjIzLTk3ZjUtNDYwNi1iM2I2LTJhZjRjMTFkNmE0ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyMVQwOTIyMTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mYmZlMzY0ZDAzZGM4OTdmYTVhMzM0YzE5N2ZkYTU3NjIzZjFmOWUzZWEzNWZkMWI5YjUzNzhlMDY5ODBkODI3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.q73NK29uMcP9pcJkVDJ3WJ20COBRP2WtnfKsVIh9G94)
> 
> That's because ECB is not taken into account during compilation phase. It does not fix the failing tests in `src/lib`

Sorry I dont understand what do you mean "ECB is not taken into account". You mean because Botan_ecb.cpp is missing in CMakelists.txt? The test was performed with dev branch and not fix-botan-ecb

zuiderkwast commented on src/lib/test/DeriveTests.cpp:
```diff
@@ -703,13 +704,31 @@ void DeriveTests::symDerive(CK_SESSION_HANDLE hSession, CK_OBJECT_HANDLE hKey, C
 			mechEncrypt.mechanism = CKM_AES_ECB;
+			#ifdef WITH_BOTAN
+			if (mechType == CKM_AES_CBC_ENCRYPT_DATA) {
+				CK_BYTE iv[16] = { 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16 };
```
I don't know what this is for, but if it's meant to be a continuous sequence from 0x1 to 0x16, then it's wrong. The numbers between 0xa and 0xf are missing.

antoinelochet replied:
I am still investigating why this test does not work on the Encrypt/Decrypt step.

zuiderkwast replied:
Is the just arbitrary data for testing? So it's not as simple as this?

```suggestion
				CK_BYTE iv[16] = { 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10 };
```

zuiderkwast replied:
Ok, I don't know. I just saw something that looked like a mistake. Ignore me.

Now, the other PR is merged.

antoinelochet replied:
I don't know either. It is a very strange behavior but Encrypt/Decrypt tests work so that must be something I miss.

antoinelochet commented:
I will close this PR since https://github.com/softhsm/SoftHSMv2/pull/771 does a better job than this.

antoinelochet commented:
#771 is merged
